File: labs/control-ui/api-shadow/ipsec_ikev2_state_machine.ts
export const UiIkeState = {
  Init: 'INIT',
  SaInitSent: 'SA_INIT_SENT',
  SaInitRecv: 'SA_INIT_RECV',
  AuthSent: 'AUTH_SENT',
  Established: 'ESTABLISHED',
  Closed: 'CLOSED',
} as const;
export type UiIkeState = (typeof UiIkeState)[keyof typeof UiIkeState];

export class IpsecIkev2StateMachine {
  public state: UiIkeState = UiIkeState.Init;
  public initiatorSpi: Uint8Array = new Uint8Array(8);
  public responderSpi: Uint8Array = new Uint8Array(8);
  private messageId: number = 0;
  public sharedSecret: Uint8Array;
  constructor(sharedSecret: Uint8Array) {
    this.sharedSecret = sharedSecret;
    if (sharedSecret.length < 16) {
      throw new Error('Shared secret must be at least 16 bytes');
    }
    for (let i = 0; i < 8; i++) {
      this.initiatorSpi[i] = Math.floor(Math.random() * 256);
    }
  }

  buildSaInitRequest(nonce: Uint8Array): Uint8Array {
    if (this.state !== UiIkeState.Init) {
      throw new Error('Cannot build SA_INIT request from non-INIT state');
    }

    const totalLen = 28 + nonce.length;
    const frame = new Uint8Array(totalLen);
    frame.set(this.initiatorSpi, 0);
    // responder SPI = 0 at offset 8
    frame[16] = 33; // Next payload: SA
    frame[17] = 0x20; // Version 2.0
    frame[18] = 34; // IKE_SA_INIT
    frame[19] = 0x08; // Initiator flag

    const view = new DataView(frame.buffer, frame.byteOffset, frame.byteLength);
    view.setUint32(20, this.messageId++);
    view.setUint32(24, totalLen);
    frame.set(nonce, 28);

    this.state = UiIkeState.SaInitSent;
    return frame;
  }

  processSaInitResponse(resp: Uint8Array): boolean {
    if (this.state !== UiIkeState.SaInitSent || resp.length < 28) return false;

    // Check initiator SPI matches
    for (let i = 0; i < 8; i++) {
      if (resp[i] !== this.initiatorSpi[i]) return false;
    }

    // Capture responder SPI
    this.responderSpi.set(resp.slice(8, 16));
    this.state = UiIkeState.SaInitRecv;
    return true;
  }

  transitionToAuth(): boolean {
    if (this.state !== UiIkeState.SaInitRecv) return false;
    this.state = UiIkeState.AuthSent;
    return true;
  }

  finalizeEstablished(): boolean {
    if (this.state !== UiIkeState.AuthSent) return false;
    this.state = UiIkeState.Established;
    return true;
  }

  close(): void {
    this.state = UiIkeState.Closed;
  }
}

File: labs/mobile/donor-archive/tunnel/IpsecIkev2StateMachine.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer
import java.security.SecureRandom

enum class AndroidIkeState {
    INIT,
    SA_INIT_SENT,
    SA_INIT_RECV,
    AUTH_SENT,
    ESTABLISHED,
    CLOSED
}

class IpsecIkev2StateMachine(val sharedSecret: ByteArray) {
    var state: AndroidIkeState = AndroidIkeState.INIT
        private set

    val initiatorSpi: Long
    var responderSpi: Long = 0
        private set
    private var messageId: Int = 0

    init {
        require(sharedSecret.size >= 16) { "Shared secret must be at least 16 bytes" }
        initiatorSpi = SecureRandom().nextLong()
    }

    fun buildSaInitRequest(nonce: ByteArray): ByteArray {
        check(state == AndroidIkeState.INIT) { "Cannot send SA_INIT from non-INIT state" }

        val totalLen = 28 + nonce.size
        val buf = ByteBuffer.allocate(totalLen)
        buf.putLong(initiatorSpi)
        buf.putLong(0L) // responder SPI is 0 in request
        buf.put(33.toByte()) // Next Payload: SA
        buf.put(0x20.toByte()) // IKEv2 version
        buf.put(34.toByte()) // Exchange Type: IKE_SA_INIT
        buf.put(0x08.toByte()) // Flags: Initiator
        buf.putInt(messageId++)
        buf.putInt(totalLen)
        buf.put(nonce)

        state = AndroidIkeState.SA_INIT_SENT
        return buf.array()
    }

    fun processSaInitResponse(resp: ByteArray): Boolean {
        if (state != AndroidIkeState.SA_INIT_SENT || resp.size < 28) return false

        val buf = ByteBuffer.wrap(resp)
        val initSpi = buf.getLong()
        if (initSpi != initiatorSpi) return false

        val respSpi = buf.getLong()
        if (respSpi == 0L) return false

        responderSpi = respSpi
        state = AndroidIkeState.SA_INIT_RECV
        return true
    }

    fun transitionToAuth(): Boolean {
        if (state != AndroidIkeState.SA_INIT_RECV) return false
        state = AndroidIkeState.AUTH_SENT
        return true
    }

    fun finalizeEstablished(): Boolean {
        if (state != AndroidIkeState.AUTH_SENT) return false
        state = AndroidIkeState.ESTABLISHED
        return true
    }

    fun close() {
        state = AndroidIkeState.CLOSED
    }
}
